### DISCUSSION TITLE
Support for NAT traversal

### DISCUSSION THREAD

**Discussion by lostmsu** [yggdrasil-network/yggdrasil-go#1140] [Q&A]
I am considering using Yggdrasil to add globally accessible IPv6 addresses to nodes in my p2p cloud gaming project: https://borg.games/

Basically the idea is that nodes will connect to the public Yggdrasil overlay and that will allow games running on them to connect to each other directly by IP address or by using a custom DNS server.

The issue is most of the nodes are behind various NATs and sometimes UPnP and other port forwarding mechanisms are not available.

Is there a built-in NAT traversal mechanism in Yggdrasil? Since QUIC is supported, it should in theory be possible to use UDP hole punching.

P.S. for IPv4 I am using https://github.com/slackhq/nebula/ which actually does have one in a form of "lighthouses"

**Comment by NikitaPuzyryov** [comment]
The question of NATs doesn't really come into play if you're not trying to make your players to be public nodes or some such. As long as your nodes/players have an outbound connection to the network (public node or other node that can reach the global network)—they can reach each other “directly” right away, using their keys (if you have a custom ygg integration, like yggmail) or their ygg IPs.

**Reply by lostmsu** [reply]
Sorry, maybe I was not clear. Without Yggdrasil I have:

Node A (private IPv4) --- NAT --- Public Internet --- NAT --- Node B (private IPv4)

In that setup Node A and Node B can not connect to each other using regular TCP/UDP. I want Node A and Node B to have IPv6 addresses through which they could reach each other. I am considering Yggdrasil for that.

I think what you are suggesting is:

Node A --- NAT --- Public Internet --- Yggdrasil public Node --- Public Internet --- NAT --- Node B

Where I use existing or run my own Yggdrasil public node.

That is not a good option, because it increases latency by adding an extra hop, and the public node may get starved by my nodes because they exchange fat video streams.

So what I want is the original

Node A  --- NAT --- Public Internet --- NAT --- Node B

Where Yggdrasil instances on A and B connect directly to each other using https://en.wikipedia.org/wiki/UDP_hole_punching or a similar protocol. E.g. either Yggdrasil needs to be able to do hole punching, or it needs to support a SOCKS 5 proxy (I have SOCKS implementation that does hole punching) that would allow it to connect to Yggdrasil running on Node B.

The reason I need this is that pretty much none of the games support SOCKS 5, so I need a virtual network they could use instead that would run over SOCKS 5 (or some other mechanism).

**Reply by NikitaPuzyryov** [reply]
Makes sense! First of all, any form of hole-punching, or ways to “discover” other distant nodes is outside of the scope of this project. Second, I'm not quite sure where yggdrasil's strengths (lightweight routing/pathfinding) come into play in your scenario. You would probably be better off by employing Tailscale's holepunching expertise, their client is [open-source](https://github.com/tailscale/tailscale), as is the unofficial control server, [headscale](https://github.com/juanfont/headscale) (developers are employed by ts).

**Reply by lostmsu** [reply]
On hole punching: if Yggdrasil would support SOCKS 5 proxy to connect to another Yggdrasil node, it would solve the hole punching for me, because I already have suitable SOCKS 5 proxy. I am kinda fond of the way SSH does it where you can just tell it a command to run instead of the host:port pair, and the command must establish connection. SSH later talks to the connection via command's stdin and stdout.

On Tailscale: AFAIK it does not support large overlay networks. The open source control server explicitly says small. Problem is when nodes for some reason can't connect directly I want them to be able to connect via Node 3 that is also part of the network. Yggdrasil does it, but Tailscale not.
